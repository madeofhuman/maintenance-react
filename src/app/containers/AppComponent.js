import React from 'react';
import { BrowserRouter, Switch, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import Homepage from './HomepageContainer';
import NotFound from './404Component';
import store from '../store/store';

const App = () => (
  <Provider store={store}>
    <BrowserRouter>
      <Switch>
        <Route path="/" exact component={Homepage} />
        <Route component={NotFound} />
      </Switch>
    </BrowserRouter>
  </Provider>
);

export default App;
