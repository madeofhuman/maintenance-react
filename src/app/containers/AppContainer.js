import React from 'react';
import { BrowserRouter, Switch, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import Homepage from './HomepageContainer';
import Dashboard from './DashboardContainer';
import NotFound from './NotFoundContainer';
import ViewRequest from './ViewRequestContainer';
import store from '../store/store';
import PrivateRoute from '../components/PrivateRouteComponent';

export const App = () => (
  <Provider store={store}>
    <BrowserRouter>
      <Switch>
        <Route path="/" exact component={Homepage} />
        <PrivateRoute exact path="/dashboard" component={Dashboard} />
        <PrivateRoute exact path="/admin" component={Dashboard} />
        <PrivateRoute exact path="/view/:requestId" component={ViewRequest} />
        <Route component={NotFound} />
      </Switch>
    </BrowserRouter>
  </Provider>
);

export default App;
