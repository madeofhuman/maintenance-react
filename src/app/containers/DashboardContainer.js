import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Navbar from '../components/NavbarComponent';
import getAllRequests from '../actions/requestActions';
import Loader from '../components/LoaderComponent';
import RequestTable from '../components/RequestTableComponent';
import spannerScrewdriver from '../assets/img/spanner-screwdriver.svg';
import Input from '../components/InputComponent';
import '../assets/css/dashboard.css';

class Dashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  componentDidMount = () => {
    const {
      authenticated, getRequests, history, user,
    } = this.props;
    if (!authenticated && user === undefined) {
      history.push('/');
    }
    if (user.role !== 'admin') {
      history.push('/dashboard');
    }
    if (user.role === 'admin') {
      history.push('/admin');
    }
    getRequests();
  }

  render() {
    const {
      loading, requests, user, message,
    } = this.props;
    return (
      <React.Fragment>
        <Navbar profileIconVisibility="" />
        <div className="body" style={{ backgroundImage: `url(${spannerScrewdriver})` }}>
          <div className="wrapper">
            <div className="left white">
              <span id="username">Welcome, {user.firstName}</span>
            </div>
            <div id="table-sort" className="right white">
              <span id="sort-group">
                <span>Sort by: </span>
                <span>
                  <select className="dropdown" id="request-status-dropdown">
                    <option value="all">All</option>
                    <option value="pending">In-Review</option>
                    <option value="approved">Pending</option>
                    <option value="ongoing">Disapproved</option>
                    <option value="resolved">Resolved</option>
                  </select>
                </span>
                &nbsp;
              </span>
              { user.role !== 'admin' ? (
                <span>
                  <a href="/new">
                    <Input id="new-request-btn" className="button" type="button" value="New Request" />
                  </a>
                </span>)
                : false }
            </div>
            <br />
            <br />
            {requests ? <RequestTable requests={requests} message={message} /> : false}
            <span className="table-nav right hidden">
              <Input type="button" value="<" className="button" />
              <Input type="button" value=">" className="button" />
            </span>
          </div>
          <br />
          <div className="divider" />
        </div>
        {loading ? <Loader /> : false}
      </React.Fragment>
    );
  }
}

Dashboard.defaultProps = {
  user: undefined,
};

Dashboard.propTypes = {
  history: PropTypes.shape({}).isRequired,
  authenticated: PropTypes.bool.isRequired,
  loading: PropTypes.bool.isRequired,
  requests: PropTypes.arrayOf(PropTypes.object).isRequired,
  message: PropTypes.string.isRequired,
  user: PropTypes.shape({}),
  getRequests: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  authenticated: state.auth.authenticated,
  loading: state.common.loading,
  history: ownProps.history,
  requests: state.requests.requests,
  message: state.requests.message,
  user: state.auth.user,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  getRequests: () => (getAllRequests.getAllRequests()),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);
