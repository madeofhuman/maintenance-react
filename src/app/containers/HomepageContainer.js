import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Navbar from '../components/NavbarComponent';
import '../assets/css/home.css';
import wrenchHammer from '../assets/img/wrench-hammer.svg';
import backgroundImage from '../assets/img/gears.svg';
import Modal from '../components/ModalComponent';
import LoginForm from './LoginContainer';
import SignupForm from './SignupContainer';

export class Homepage extends Component {
  state = {
    show: false,
    modalContent: '',
  };

  showModal = (event) => {
    event.preventDefault();
    this.setState({
      show: true,
      modalContent: event.target.name,
    });
  };

  hideModal = () => {
    this.setState({
      show: false,
      modalContent: '',
    });
  };

  componentDidMount = () => {
    const { authenticated, history, role } = this.props;
    if (authenticated) {
      if (role === 'user') {
        history.push('/dashboard');
      }
      if (role === 'admin') {
        history.push('admin');
      }
      if (!role) {
        return true;
      }
    }
  }

  render() {
    const { show, modalContent } = this.state;
    const { history } = this.props;
    return (
      <React.Fragment>
        <Modal show={show} handleClose={this.hideModal}>
          { modalContent === 'sign-in' ? <LoginForm handleClose={this.hideModal} history={history} /> : <SignupForm handleClose={this.hideModal} /> }
        </Modal>
        <Navbar />
        <div className="">
          <div className="body" style={{ backgroundImage: `url(${backgroundImage})` }}>
            <div className="article left">
              <span className="icon">
                <img src={wrenchHammer} alt="icon" />
              </span>
              <span className="text right white">
                <h2>
                  Nobody should have to live with broken things.
                </h2>
                <br />
                <p>
                  We want to help you fix the things you love, and make sure the things that matter
                  most to you, do not fall to pieces, so you can actually enjoy using them.
                  <br /><br />
                  We take great pride in our work and our ability to exceed customer expectation.
                  You, the customer, are our greatest resource and we will treat you accordingly.
                </p>
              </span>
            </div>
            <div id="login-links" className="right">
              <span className="left link-btn">
                <a title="Sign in" name="sign-in" className="white" id="sign-in-link" onClick={this.showModal}>Sign in</a>
              </span>
              <span className="pipe white">
                |
              </span>
              <span className=" link-btn">
                <a title="Sign up" name="sign-up" className="orange" id="sign-up-link" onClick={this.showModal}>Sign up</a>
              </span>
            </div>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

Homepage.defaultProps = {
  role: null,
};

Homepage.propTypes = {
  history: PropTypes.shape({}).isRequired,
  authenticated: PropTypes.bool.isRequired,
  role: PropTypes.string,
};

const mapStateToProps = (state, ownProps) => ({
  authenticated: state.auth.authenticated,
  history: ownProps.history,
  role: state.auth.user !== null ? state.auth.user.role : null,
});

export default connect(mapStateToProps, null)(Homepage);
