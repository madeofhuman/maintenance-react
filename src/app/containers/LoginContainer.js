import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Input from '../components/common/Input';
import authenticate from '../actions/authActions';
import Loader from '../components/common/Loader';

class LoginForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
    };
  }

  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  };

  handleSubmit = (event) => {
    event.preventDefault();
    const {
      email, password,
    } = this.state;
    const { login, history } = this.props;
    const user = {
      email, password,
    };
    return login(user, history, 'login');
  };

  render() {
    const {
      email, password,
    } = this.state;
    const { handleClose, loading } = this.props;
    return (
      <React.Fragment>
        <div id="sign-in-form" className="form">
          <form id="login-form" onSubmit={this.handleSubmit}>
            <span id="sign-up-form-close-btn" className="close-btn grey right" onClick={handleClose}>x</span>
            <p className="output white" />
            <Input
              type="text"
              name="email"
              id="login-email"
              label="Email"
              value={email}
              handleChange={this.handleChange}
              isRequired="required"
            />
            <br /><br />
            <Input
              type="password"
              name="password"
              id="login-password"
              label="Password"
              value={password}
              handleChange={this.handleChange}
              isRequired="required"
            />
            <br /><br />
            <Input
              type="submit"
              value="Log in"
              className="white auth-submit-btn"
              id="signin-btn"
            />
          </form>
        </div>
        { loading ? <Loader /> : '' }
      </React.Fragment>
    );
  }
}

LoginForm.propTypes = {
  handleClose: PropTypes.func.isRequired,
  history: PropTypes.shape({}).isRequired,
  login: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  loading: state.common.loading,
  history: ownProps.history,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  login: (user, history, route) => (authenticate.authenticate(user, history, route)),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);
