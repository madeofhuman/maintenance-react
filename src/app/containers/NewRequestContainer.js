import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import toastr from 'toastr';
import RequestForm from '../components/RequestForm';
import requestActions from '../actions/requestActions';

/**
 * Manages the state and actions of the New Request Component
 * @class
 * @extends React.Component
 */
export class NewRequest extends Component {
  constructor(props) {
    super(props);
    this.state = {
      type: '',
      item: '',
      model: '',
      detail: '',
    };
  }

  /**
   * Updates the value of the item in the form state with the value of their bound
   * DOM elements when the DOM elements change
   */
  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  /**
   * Validates the user input and calls the createRequest action dispatcher
   * to create the request if the input is valid, or toast an error otherwise.
   */
  handleSubmit = (event) => {
    event.preventDefault();
    const {
      createRequest, getAllRequests,
    } = this.props;
    const {
      type, item, model, detail,
    } = this.state;
    const requestData = {
      type, item, model, detail,
    };
    if (requestData.type.trim() === '') return toastr.error('Type cannot be empty.');
    if (requestData.item.trim() === '') return toastr.error('Item cannot be empty.');
    if (requestData.detail.trim() === '') return toastr.error('Description cannot be empty.');
    if (requestData.model.trim() === '') requestData.model = 'N/A';
    createRequest(requestData);
    getAllRequests();
  }

  /**
   * Renders the New Request Form component on a node in the DOM
   */
  render() {
    const { handleClose } = this.props;
    return (
      <RequestForm
        handleSubmit={this.handleSubmit}
        handleClose={handleClose}
        handleChange={this.handleChange}
      />
    );
  }
}

NewRequest.defaultProps = {
  history: undefined,
};

NewRequest.propTypes = {
  history: PropTypes.shape({}),
  createRequest: PropTypes.func.isRequired,
  getAllRequests: PropTypes.func.isRequired,
  handleClose: PropTypes.func.isRequired,
};

/**
 * Add specified items in the global store as props to the component
 * @param {object} state the global store
 * @param {object} ownProps the component specific props
 */
export const mapStateToProps = (state, ownProps) => ({
  history: ownProps.history,
});

/**
 * Add specified action creators as props to the component
 * @param {function} dispatch - dispatch the specified action.
 */
export const mapDispatchToProps = dispatch => bindActionCreators({
  createRequest: requestData => (requestActions.createRequest(requestData)),
  getAllRequests: () => (requestActions.getAllRequests()),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(NewRequest);
