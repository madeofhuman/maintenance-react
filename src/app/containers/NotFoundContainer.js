import React from 'react';
import Navbar from '../components/NavbarComponent';
import '../assets/css/404.css';

export const NotFound = () => (
  <React.Fragment>
    <div className="wrapper">
      <Navbar />
      <div className="body" style={{ backgroundImage: "url('../assets/img/gears.svg')" }}>
        <div className="wrapper white">
          <h1 className="center" id="four-o-four-message">
            <span role="img" aria-label="sad face">ðŸ˜¢</span><br />Nothing exists here.
          </h1>
          <em className="center" id="back-btn" onClick={() => (window.history.back())}>&lt;&nbsp;&nbsp;Go Back</em>
        </div>
      </div>
    </div>
  </React.Fragment>
);

export default NotFound;
