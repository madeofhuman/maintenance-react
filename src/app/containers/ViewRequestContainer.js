import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Modal from 'react-modal';
import Navbar from '../components/NavbarComponent';
import getAllRequests from '../actions/requestActions';
import '../assets/css/request.css';
import cloudGearSpanner from '../assets/img/cloud-gear-spanner.svg';
import RequestCard from '../components/RequestCardComponent';
import Loader from '../components/LoaderComponent';
import UpdateRequest from './UpdateRequestContainer';

/**
 * Manages the state and actions of the View Request Component
 * @class
 * @extends React.Component
 */
export class ViewRequest extends Component {
  constructor(props) {
    super(props);
    this.state = {
      show: false,
    };
  }

  /**
   * Fetches the request when the component mounts
   */
  componentDidMount = async () => {
    const { match, getRequest } = this.props;
    const { requestId } = match.params;
    await getRequest(requestId);
  }

  /**
   * Mounts the modal with the edit request form
   */
  showModal = () => {
    this.setState({
      show: true,
    });
  };

  /**
   * Unmounts the modal with the edit request form
   */
  hideModal = () => {
    this.setState({
      show: false,
    });
  };

  /**
   * Delete the request the id of which is gotten from the url param
   * and redirect to the dashboard
   */
  deleteRequest = () => {
    const { match, deleteRequest, history } = this.props;
    const { requestId } = match.params;
    deleteRequest(requestId);
    history.push('/');
  }

  /**
   * Approve the request the id of which is gotten from the url param
   */
  approveRequest = () => {
    const { match, approveRequest } = this.props;
    const { requestId } = match.params;
    approveRequest(requestId);
  }

  /**
   * Disapprove the request the id of which is gotten from the url param
   */
  disapproveRequest = () => {
    const { match, disapproveRequest } = this.props;
    const { requestId } = match.params;
    disapproveRequest(requestId);
  }

  /**
   * Resolve the request the id of which is gotten from the url param
   */
  resolveRequest = () => {
    const { match, resolveRequest } = this.props;
    const { requestId } = match.params;
    resolveRequest(requestId);
  }

  /**
   * Renders the View Request component on a node in the DOM
   */
  render() {
    const {
      request, loading, user, history, match,
    } = this.props;
    const { show } = this.state;
    if (loading) {
      return <Loader />;
    }
    return (
      <React.Fragment>
        <Modal
          isOpen={show}
          handleClose={this.hideModal}
          ariaHideApp={false}
          overlayClassName="modal-overlay"
          className="modal-content"
        >
          <UpdateRequest
            handleClose={this.hideModal}
            initialValues={request}
            history={history}
            match={match}
          />
        </Modal>
        <Navbar />
        <div className="body" style={{ backgroundImage: `url(${cloudGearSpanner})` }}>
          <p className="output white center" id="output" />
          <div className="wrapper" id="wrapper">
            { typeof request !== 'array' ? (
              <RequestCard
                request={request}
                role={user.role}
                requestActions={{
                  deleteRequest: this.deleteRequest,
                  approveRequest: this.approveRequest,
                  disapproveRequest: this.disapproveRequest,
                  resolveRequest: this.resolveRequest,
                }}
                showModal={this.showModal}
              />)
              : <p className="white">Request does not exist</p> }
          </div>
        </div>
      </React.Fragment>
    );
  }
}

ViewRequest.propTypes = {
  request: PropTypes.shape({}).isRequired,
  getRequest: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  user: PropTypes.shape({}).isRequired,
  match: PropTypes.shape({}).isRequired,
  deleteRequest: PropTypes.func.isRequired,
  approveRequest: PropTypes.func.isRequired,
  disapproveRequest: PropTypes.func.isRequired,
  resolveRequest: PropTypes.func.isRequired,
  history: PropTypes.shape({}).isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  loading: state.common.loading,
  request: state.requests.request,
  match: ownProps.match,
  user: state.auth.user,
  history: ownProps.history,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  getRequest: requestData => (getAllRequests.getSingleRequest(requestData)),
  deleteRequest: requestId => (getAllRequests.deleteRequest(requestId)),
  approveRequest: requestId => (getAllRequests.approveRequest(requestId)),
  disapproveRequest: requestId => (getAllRequests.disapproveRequest(requestId)),
  resolveRequest: requestId => (getAllRequests.resolveRequest(requestId)),
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(ViewRequest);
