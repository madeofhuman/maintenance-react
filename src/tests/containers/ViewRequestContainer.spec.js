import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { ViewRequest } from '../../app/containers/ViewRequestContainer';

Enzyme.configure({ adapter: new Adapter() });

const props = {
  request: {
    id: 1,
    type: 'repair',
    item: 'Macbook Pro',
    model: '2015',
    detail: "it's stuck in a boot loop",
    owner: 'josephodina@gmail.com',
  },
  getRequest: jest.fn(),
  loading: false,
  user: {
    firstName: 'Chukwuka',
    lastName: 'Odina',
    email: 'josephodina@gmail.com',
    role: 'user',
  },
  match: {
    params: {
      requestId: 1,
    },
  },
  deleteRequest: jest.fn(),
  approveRequest: jest.fn(),
  disapproveRequest: jest.fn(),
  resolveRequest: jest.fn(),
};

describe('View Request Component', () => {
  test('renders', () => {
    const wrapper = shallow(<ViewRequest {...props} />);
    expect(wrapper.exists()).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
