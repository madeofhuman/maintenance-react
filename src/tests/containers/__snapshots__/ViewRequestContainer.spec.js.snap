// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Request Component renders 1`] = `
<Fragment>
  <Modal
    handleClose={[Function]}
    show={false}
  >
    <Connect(RequestForm)
      handleClose={[Function]}
      initialValues={
        Object {
          "detail": "it's stuck in a boot loop",
          "id": 1,
          "item": "Macbook Pro",
          "model": "2015",
          "owner": "josephodina@gmail.com",
          "type": "repair",
        }
      }
    />
  </Modal>
  <Connect(Navbar) />
  <div
    className="body"
    style={
      Object {
        "backgroundImage": "url(null)",
      }
    }
  >
    <p
      className="output white center"
      id="output"
    />
    <div
      className="wrapper"
      id="wrapper"
    >
      <RequestCard
        request={
          Object {
            "detail": "it's stuck in a boot loop",
            "id": 1,
            "item": "Macbook Pro",
            "model": "2015",
            "owner": "josephodina@gmail.com",
            "type": "repair",
          }
        }
        requestActions={
          Object {
            "approveRequest": [Function],
            "deleteRequest": [Function],
            "disapproveRequest": [Function],
            "resolveRequest": [Function],
          }
        }
        role="user"
        showModal={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
